File: backend-ts/src/services/ProductService.ts
import { Database } from '../database/Database';
import { Product, CreateProductRequest, UpdateProductRequest, StockUpdateRequest } from '../types/Product';

export class ProductService {
    constructor(private database: Database) { }

    async getAllProducts(): Promise<Product[]> {
        console.log('Fetching all products');
        return await this.database.getAllProducts();
    }

    async getProductById(id: number): Promise<Product> {
        console.log(`Fetching product with ID: ${id}`);
        const product = await this.database.getProductById(id);
        if (!product) {
            throw new Error(`Product not found with ID: ${id}`);
        }
        return product;
    }

    async createProduct(request: CreateProductRequest): Promise<Product> {
        console.log(`Creating product with name: ${request.name}`);

        // Validate input
        if (!request.name || request.name.trim().length === 0) {
            throw new Error('Product name is required');
        }

        if (request.stockQuantity < 0) {
            throw new Error('Stock quantity cannot be negative');
        }

        if (request.lowStockThreshold !== undefined && request.lowStockThreshold < 0) {
            throw new Error('Low stock threshold cannot be negative');
        }

        const productData = {
            name: request.name.trim(),
            description: request.description?.trim(),
            stockQuantity: request.stockQuantity,
            lowStockThreshold: request.lowStockThreshold || 10
        };

        const product = await this.database.createProduct(productData);
        console.log(`Product created with ID: ${product.id}`);
        return product;
    }

    async updateProduct(id: number, request: UpdateProductRequest): Promise<Product> {
        console.log(`Updating product with ID: ${id}`);

        // Validate input
        if (request.name !== undefined && (!request.name || request.name.trim().length === 0)) {
            throw new Error('Product name cannot be empty');
        }

        if (request.stockQuantity !== undefined && request.stockQuantity < 0) {
            throw new Error('Stock quantity cannot be negative');
        }

        if (request.lowStockThreshold !== undefined && request.lowStockThreshold < 0) {
            throw new Error('Low stock threshold cannot be negative');
        }

        const updates: Partial<Omit<Product, 'id' | 'lowStock'>> = {};

        if (request.name !== undefined) {
            updates.name = request.name.trim();
        }
        if (request.description !== undefined) {
            updates.description = request.description?.trim();
        }
        if (request.stockQuantity !== undefined) {
            updates.stockQuantity = request.stockQuantity;
        }
        if (request.lowStockThreshold !== undefined) {
            updates.lowStockThreshold = request.lowStockThreshold;
        }

        const product = await this.database.updateProduct(id, updates);
        if (!product) {
            throw new Error(`Product not found with ID: ${id}`);
        }

        console.log(`Product updated with ID: ${id}`);
        return product;
    }

    async deleteProduct(id: number): Promise<void> {
        console.log(`Deleting product with ID: ${id}`);
        const deleted = await this.database.deleteProduct(id);
        if (!deleted) {
            throw new Error(`Product not found with ID: ${id}`);
        }
        console.log(`Product deleted with ID: ${id}`);
    }

    async increaseStock(id: number, request: StockUpdateRequest): Promise<Product> {
        console.log(`Increasing stock for product ID: ${id} by quantity: ${request.quantity}`);

        if (request.quantity <= 0) {
            throw new Error('Quantity must be greater than 0');
        }

        const product = await this.getProductById(id);
        const newStockQuantity = product.stockQuantity + request.quantity;

        const updatedProduct = await this.database.updateProduct(id, { stockQuantity: newStockQuantity });
        if (!updatedProduct) {
            throw new Error(`Product not found with ID: ${id}`);
        }

        console.log(`Stock increased for product ID: ${id}. New stock: ${updatedProduct.stockQuantity}`);
        return updatedProduct;
    }

    async decreaseStock(id: number, request: StockUpdateRequest): Promise<Product> {
        console.log(`Decreasing stock for product ID: ${id} by quantity: ${request.quantity}`);

        if (request.quantity <= 0) {
            throw new Error('Quantity must be greater than 0');
        }

        const product = await this.getProductById(id);

        if (product.stockQuantity < request.quantity) {
            const errorMessage = `Insufficient stock. Available: ${product.stockQuantity}, Requested: ${request.quantity}`;
            console.error(`Insufficient stock for product ID: ${id}. ${errorMessage}`);
            throw new Error(errorMessage);
        }

        const newStockQuantity = product.stockQuantity - request.quantity;
        const updatedProduct = await this.database.updateProduct(id, { stockQuantity: newStockQuantity });
        if (!updatedProduct) {
            throw new Error(`Product not found with ID: ${id}`);
        }

        console.log(`Stock decreased for product ID: ${id}. New stock: ${updatedProduct.stockQuantity}`);
        return updatedProduct;
    }

    async getLowStockProducts(): Promise<Product[]> {
        console.log('Fetching products with low stock');
        return await this.database.getLowStockProducts();
    }
}

File: backend-java/ase-challenge/src/main/java/com/verto/ase_challenge/service/ProductService.java
package com.verto.ase_challenge.service;

import com.verto.ase_challenge.dto.ProductRequest;
import com.verto.ase_challenge.dto.ProductResponse;
import com.verto.ase_challenge.entity.Product;
import com.verto.ase_challenge.exception.InsufficientStockException;
import com.verto.ase_challenge.exception.ProductNotFoundException;
import com.verto.ase_challenge.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class ProductService {
    
    private final ProductRepository productRepository;
    
    public ProductResponse createProduct(ProductRequest request) {
        log.info("Creating product with name: {}", request.getName());
        
        Product product = new Product(
            request.getName(),
            request.getDescription(),
            request.getStockQuantity(),
            request.getLowStockThreshold()
        );
        
        Product savedProduct = productRepository.save(product);
        log.info("Product created with ID: {}", savedProduct.getId());
        
        return ProductResponse.fromEntity(savedProduct);
    }
    
    @Transactional(readOnly = true)
    public ProductResponse getProductById(Long id) {
        log.info("Fetching product with ID: {}", id);
        Product product = findProductById(id);
        return ProductResponse.fromEntity(product);
    }
    
    @Transactional(readOnly = true)
    public List<ProductResponse> getAllProducts() {
        log.info("Fetching all products");
        return productRepository.findAll()
            .stream()
            .map(ProductResponse::fromEntity)
            .collect(Collectors.toList());
    }
    
    public ProductResponse updateProduct(Long id, ProductRequest request) {
        log.info("Updating product with ID: {}", id);
        Product product = findProductById(id);
        
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setStockQuantity(request.getStockQuantity());
        product.setLowStockThreshold(request.getLowStockThreshold());
        
        Product updatedProduct = productRepository.save(product);
        log.info("Product updated with ID: {}", updatedProduct.getId());
        
        return ProductResponse.fromEntity(updatedProduct);
    }
    
    public void deleteProduct(Long id) {
        log.info("Deleting product with ID: {}", id);
        Product product = findProductById(id);
        productRepository.delete(product);
        log.info("Product deleted with ID: {}", id);
    }
    
    public ProductResponse increaseStock(Long id, Integer quantity) {
        log.info("Increasing stock for product ID: {} by quantity: {}", id, quantity);
        Product product = findProductById(id);
        
        int newStock = product.getStockQuantity() + quantity;
        product.setStockQuantity(newStock);
        
        Product updatedProduct = productRepository.save(product);
        log.info("Stock increased for product ID: {}. New stock: {}", id, newStock);
        
        return ProductResponse.fromEntity(updatedProduct);
    }
    
    public ProductResponse decreaseStock(Long id, Integer quantity) {
        log.info("Decreasing stock for product ID: {} by quantity: {}", id, quantity);
        Product product = findProductById(id);
        
        if (product.getStockQuantity() < quantity) {
            String message = String.format(
                "Insufficient stock. Available: %d, Requested: %d", 
                product.getStockQuantity(), 
                quantity
            );
            log.error("Insufficient stock for product ID: {}. {}", id, message);
            throw new InsufficientStockException(message);
        }
        
        int newStock = product.getStockQuantity() - quantity;
        product.setStockQuantity(newStock);
        
        Product updatedProduct = productRepository.save(product);
        log.info("Stock decreased for product ID: {}. New stock: {}", id, newStock);
        
        return ProductResponse.fromEntity(updatedProduct);
    }
    
    @Transactional(readOnly = true)
    public List<ProductResponse> getLowStockProducts() {
        log.info("Fetching products with low stock");
        return productRepository.findLowStockProducts()
            .stream()
            .map(ProductResponse::fromEntity)
            .collect(Collectors.toList());
    }
    
    private Product findProductById(Long id) {
        return productRepository.findById(id)
            .orElseThrow(() -> new ProductNotFoundException("Product not found with ID: " + id));
    }
}